### PR TITLE
[inductor] don't cache non-static content

### PR DESCRIPTION
**Pull request by shunting314**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #100331
* __->__ #106502

I happened to find that inductor may cache stale inner_fn_str and ReadWrites object in a ComputedBuffer when I work on looping ordering.

Let's say we have producer buffer buf0 and consumer buffer buf1. Before we call GraphLowering.finalize, the layout for buf0 may be a FlexibleLayout. At that moment, the inner_fn_str or ReadWrites object computed for buf1 will be based on the layout of buf0 which most likely is a contiguous FlexibleLayout. And they will be cached on buf1 object (or buf1.data).

However after we call GraphLowering.finalize, we may realize it's better to give a non-contiguous layout for buf0 (e.g., if its input has non-contiguous layout or whatever reason). The layout change of buf0 should affect the inner_fn_str and ReadWrites object for buf1. But we may have cached those on buf1. The stale ReadWrites objects for buf1 may result in sub-optimal strides for buf1.

This may affect perf and I'll check the nightly runs.

Here is a dump of `nodes` in `Scheduler.__init__` before the fix as a reference: https://gist.github.com/shunting314/ed2152a08e268f5563fd55398b1392c7 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

**Review by jansel — CHANGES_REQUESTED** [review]
Thanks for finding this!

Instead of clearing the cache, I'd suggest just removing the `@cache_on_self` decorators.  If a value isn't static, we shouldn't cache on it.  Needing to clear an improperly cached value sounds error-prone.

**Comment by shunting314** [comment]
> Instead of clearing the cache, I'd suggest just removing the @cache_on_self decorators. 

I agree. But I'm a bit worried about compilation time. `Loops.inner_fn_str` is fine since it's not called a lot. But `ComputedBuffer.get_read_writes` seems to be called relatively often. But I'll run a perf experiment to verify on the dashboard.

**Comment by jansel** [comment]
> I agree. But I'm a bit worried about compilation time. `Loops.inner_fn_str` is fine since it's not called a lot. But `ComputedBuffer.get_read_writes` seems to be called relatively often. But I'll make the suggested change and run a perf experiment to verify on the dashboard.

Yeah there might be a compile time hit, though I think correctness is more important.

There is stuff we could cache safely.  I believe inner_fn_str is only used in a "length of body" heuristic.  get_read_writes() is used for "number of reads" and "names of buffers read".  None of those things would change with layout, perhaps we should just precompute all 3 of them in the constructor.

**Review by jansel — APPROVED** [review]

**Review by eellison — COMMENTED** [review]
Nice catch! Would you post the compilation numbers when they are done running on this PR ?

**Comment by shunting314** [comment]
> Would you post the compilation numbers when they are done running on this PR ?

Here is the result on the dashboard: [link](https://hud.pytorch.org/benchmark/compilers?startTime=Thu%2C%2020%20Jul%202023%2017%3A37%3A29%20GMT&stopTime=Thu%2C%2003%20Aug%202023%2017%3A37%3A29%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=shunting-avoid-stale-cache&lCommit=bdb7698f530063e712c3308b0e89d4c00079758c&rBranch=main&rCommit=a1d0db1c60b746c1c9f2500bb231c528a461eda5 )
- perf is neutral
- there is 1s geomean compilation time regress in some settings. Looks like not a big price to pay. I'll cache more 'derived' content Jason suggested above. Hopefully we can reduce the compilation time hit
